# 3D de novo assembly (3D-DNA) pipeline with AGWG-specific modifications

This is a version of the [3D-DNA pipeline (Dudchenko et al., Science, 2017)] that was used to help generate [AaegL5] genome assembly for the mosquito *Aedes aegypti*.

In addition to the scripts contained here, the Hi-C portion of the workflow for AaegL5 used tools and methods from Juicer [(Durand & Shamim et al., Cell Systems, 2016)], Juicebox [(Durand & Robinson et al., Cell Systems, 2016)] and Juicebox Assembly Tools [(Dudchenko et al., bioRxiv, 2018)]. The *in situ* Hi-C experiments were performed as described in [(Rao & Huntley et al., Cell, 2014)]. Interactive figures incorporated with this description were created using Juicebox.js software [(Robinson et al., Cell Systems, 2018)].

For the latest release of 3D-DNA see [3d-dna].

Feel free to post your questions and comments at:
[http://www.aidenlab.org/forum.html]

### Workflow, Commands and Intermediate Results
The detailed description of the workflow is hosted [here]. In brief, scaffolding and alternative haplotype removal for AaegL5 genome assembly can be broken down into 7 steps:  
- 1: Automatic preliminary scaffolding;
- 2: Curated misjoin correction based on preliminary scaffolding results;
- 3: Automatic scaffolding of misjoin-corrected contigs;
- 4: Curated polishing and chromosome splitting;
- 5: Automatic pairwise alignment of contigs that were placed in nearby positions along the pre-merge assembly
- 6: Curated arrangement of contigs into 'merge groups' based on alignment data
- 7: Automatic merging of overlapping contigs in individual merge groups

Listed below are commands and intermediate processing files associated with these steps, as well as input data (included as step 0: Draft contigs) and final assembly (included as step 8: Final NCBI submission).

##### Step 0: Draft contigs

For data associated with the draft FALCON-Unzip contigs see the following files:
- [AGWG.draft.fasta.gz](https://www.dropbox.com/s/5a864lchwv1dost/AGWG.draft.fasta.gz?dl=0) - the FALCON-Unzip output;
- [AGWG.draft.mnd.gz](https://www.dropbox.com/s/5a864lchwv1dost/AGWG.draft.fasta.gz?dl=0) - list of paired alignments of Hi-C reads (combined from three *in situ* Hi-C experiments [SRR6294665](https://www.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6294665), [SRR6294663](https://www.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6294663), [SRR6294664](https://www.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6294664)) to the draft FALCON-Unzip fasta as generated by the Juicer pipeline;
- [AGWG.draft.assembly](https://www.dropbox.com/s/du80vxvp9xsnpfv/AGWG.draft.assembly?dl=0) - *.assembly* file format, generated by 3D-DNA and Juicebox Assembly Tools, that stores the relationship between the original draft sequences and the output assembly (see below);
- [AGWG.draft.hic](https://www.dropbox.com/s/1at5l20txc0czun/AGWG.draft.hic?dl=0), [AGWG.draft_0.hic](https://www.dropbox.com/s/ipf6sagb6j3m7fy/AGWG.draft_0.hic?dl=0) and [AGWG.draft_30.hic](https://www.dropbox.com/s/h41o02t9lryvmlz/AGWG.draft_30.hic?dl=0) - Hi-C contact maps in the *.hic* file format compatible with Juicebox and Juicebox Assembly Tools thresholded at mapping quality 1, 0 and 30, respectively, generated by 3D-DNA visualization module (see below);
- [AGWG.draft_asm.scaffold_track.txt](https://www.dropbox.com/s/uhbqyxt9dhjmjaj/AGWG.draft_asm.scaffold_track.txt?dl=0) and [AGWG.draft_asm.superscaf_track.txt](https://www.dropbox.com/s/vl4xio4r19mh19d/AGWG.draft_asm.superscaf_track.txt?dl=0) - 2D annotation tracks marking down the boundaries of scaffolds and superscaffolds, respectively, as generated by 3D-DNA visualization module (see below).

Commands used to generate the files are as follows:
```sh
# prep: generate assembly file from falcon full output fasta
awk -f ${path_to_3ddna}/utils/generate-assembly-file-from-fasta.awk AGWG.draft.fasta > AGWG.draft.assembly
```
```sh
# generate a high-resolution Hi-C map to visualize the draft with default mapping quality threshold (mapq>=1)
bash ${path_to_3ddna}/visualize/run-assembly-visualizer.sh -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.draft.assembly AGWG.mnd.txt
```
```sh
# generate supplementary Hi-C maps to visualize draft with mapq>=0 and mapq>=30:
bash ${path_to_3ddna}/visualize/run-assembly-visualizer.sh -q 0 -m temp.AGWG.draft.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.draft.assembly AGWG.mnd.txt
bash ${path_to_3ddna}/visualize/run-assembly-visualizer.sh -q 30 -m temp.AGWG.draft.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.draft.assembly AGWG.mnd.txt
```
##### Step 1: Preliminary scaffolding
The preliminary scaffolding step results include the following files:
- [AGWG.0.assembly](https://www.dropbox.com/s/zdke16o5csy1xpi/AGWG.0.assembly?dl=0) - *.assembly* file generated by 3D-DNA (see below);
- [AGWG.0.hic](https://www.dropbox.com/s/kj0rcw4zkgxb8fx/AGWG.0.hic?dl=0), [AGWG.0_0.hic](https://www.dropbox.com/s/urtyabxyl3rywu1/AGWG.0_0.hic?dl=0), [AGWG.0_30.hic](https://www.dropbox.com/s/2ihmtpb2rztudt0/AGWG.0_30.hic?dl=0) - Hi-C contact maps, as generated by the 3D-DNA pipeline visualization module (see below), thresholded at mapping quality 1, 0 and 30, respectively (see below);
- [AGWG.0_asm.scaffold_track.txt](https://www.dropbox.com/s/w9jji47vhkpgf0p/AGWG.0_asm.scaffold_track.txt?dl=0) and [AGWG.0_asm.superscaf_track.txt](https://www.dropbox.com/s/wa7q9pm997lxbio/AGWG.0_asm.superscaf_track.txt?dl=0) - 2D annotation tracks for this step, as generated by 3D-DNA visualization module (see below).

Commands used to generate these files are listed below:

```sh
# prep
awk -f ${path_to_3ddna}/utils/convert-assembly-to-cprop-and-asm.awk AGWG.draft.assembly
ln -sf AGWG.draft.cprops AGWG.0.cprops
ln -sf AGWG.mnd.txt AGWG.mnd.0.txt
```
```sh
# order and orient scaffolds larger than 20000
bash ${path_to_3ddna}/scaffold/run-liger-assembler.sh -s 20000 AGWG.0.cprops AGWG.mnd.0.txt
```
```sh
# build high-resolution hic map with mapq >= 1 (default):
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.0.cprops AGWG.0.asm AGWG.mnd.0.txt
```
```sh
# build supplementary high-resolution hic maps with reads mapping quality >=0 and >=30.
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -q 0 -m temp.AGWG.0.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.0.cprops AGWG.0.asm AGWG.mnd.0.txt
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -q 30 -m temp.AGWG.0.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.0.cprops AGWG.0.asm AGWG.mnd.0.txt
```
##### Step 2: Curated misjoin correction

Misassembly correction has been performed by surveying the preliminary scaffolding results in Juicebox Assembly Tools. Regions annotated as misassembled are shared as a [2D annotation track](https://www.dropbox.com/s/oonfp9qnadmf66p/suspect_2D.at.step.0.txt?dl=0). For convenience, the data is also presented in a form of an [input 2D annotation track](https://www.dropbox.com/s/bndonjc7dkz1z6v/AGWG.edits.txt?dl=0) and a [bed track](https://www.dropbox.com/s/xviy7p20zaklt45/suspect.at.step.0.bed?dl=0) as generated by 3D-DNA (see below).

The annotated regions can be surveyed using [this interactive figure](http://bit.ly/2JwrCJb), which shows a Hi-C contact map from Step 1: Preliminary scaffolding ([AGWG.0_0.hic](https://www.dropbox.com/s/urtyabxyl3rywu1/AGWG.0_0.hic?dl=0)). The heatmap indicates the frequency of contact between pairs of loci in the assembly as measured by the Hi-C experiments, on a scale from white to red (pure white indicates no observed contacts, whereas pure red indicates a maximum number of contacts which is indicated in the top left corner of the interactive figure). 2D squares along the diagonal indicate the boundaries of input contigs as ordered and oriented in Step 1 ([AGWG.0_asm.scaffold_track.txt](https://www.dropbox.com/s/w9jji47vhkpgf0p/AGWG.0_asm.scaffold_track.txt?dl=0)). Shown on top and to the left of the map, along the X and Y axes, is a *.bed* track ([suspect.at.step.0.bed](https://www.dropbox.com/s/xviy7p20zaklt45/suspect.at.step.0.bed?dl=0)) highlighting the annotated misjoin positions: horizontal and vertical blue bars align with diagonal points inside the 2D annotation squares that are associated with a reduced number of contacts between sequences immediately upstream and downstream than would be expected in the case of correctly joined sequences. By default the figure focuses on a representative interval; by scrolling readers can explore the complete, genome-wide data and annotations.

The command used to edit the input contigs based on the annotation file is included below:
```sh
( (awk 'NR==1{print > "/dev/stderr";next}1' <(awk -f ${path_to_3ddna}/lift/lift-asm-annotations-to-input-annotations.awk AGWG.0.cprops AGWG.0.asm suspect_2D.at.step.0.txt) | sort -k 2,2n -k 3,3n)  2>&1 ) > AGWG.edits.txt
```
##### Step 3: Automatic scaffolding of misjoin-corrected contigs

The scaffolding step using misjoin-corrected contigs as input is documented with the following files:
- [AGWG.1.assembly](https://www.dropbox.com/s/fx1q1uwhlmn31hw/AGWG.1.assembly?dl=0) - *.assembly* file generated by 3D-DNA (see below);
- [AGWG.1.hic](https://www.dropbox.com/s/lwvv1gs0jpudil2/AGWG.1.hic?dl=0), [AGWG.1_0.hic](https://www.dropbox.com/s/o76oxr6x0gpwnj0/AGWG.1_0.hic?dl=0), [AGWG.1_30.hic](https://www.dropbox.com/s/spsskyllbaafw1b/AGWG.1_30.hic?dl=0) - Hi-C contact maps as generated by the 3D-DNA pipeline visualization module (see below), thresholded at mapping quality 1, 0 and 30, respectively (see below);
- [AGWG.1_asm.scaffold_track.txt](https://www.dropbox.com/s/7xndgilq6xgcbeq/AGWG.1_asm.scaffold_track.txt?dl=0) and [AGWG.1_asm.superscaf_track.txt](https://www.dropbox.com/s/3f05phfonrng0ka/AGWG.1_asm.superscaf_track.txt?dl=0) - 2D annotation tracks for this step as generated by 3D-DNA visualization module (see below).

Commands used to generate the above files are listed below:
```sh
# prep
awk -f ${path_to_3ddna}/edit/edit-cprops-according-to-annotations.awk AGWG.edits.txt AGWG.0.cprops > AGWG.1.cprops
bash ${path_to_3ddna}/edit/edit-mnd-according-to-new-cprops.sh AGWG.1.cprops AGWG.mnd.txt > AGWG.mnd.1.txt
```
```sh
# order and orient scaffolds larger than 20000
bash ${path_to_3ddna}/scaffold/run-liger-assembler.sh -s 20000 AGWG.1.cprops AGWG.mnd.1.txt
```
```sh
# build high-resolution hic map with mapq >= 1 (default):
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.1.cprops AGWG.1.asm AGWG.mnd.1.txt
```
```sh
# build supplementary high-resolution hic maps with reads mapping quality >=0 and >=30.
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -q 0 -m temp.AGWG.1.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.1.cprops AGWG.1.asm AGWG.mnd.1.txt
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -q 30 -m temp.AGWG.1.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.1.cprops AGWG.1.asm AGWG.mnd.1.txt
```
##### Step 4: Curated polishing and chromosome splitting
The assembly from step 3 was manually polished using Juicebox Assembly Tools to generate three raw chromosomal scaffolds. Shared below is the reviewed assembly file as exported by the Juicebox Assembly Tools as well as the Hi-C contact maps associated with the step:
- [AGWG.1.review.assembly](https://www.dropbox.com/s/wgyxu0lys68e6h8/AGWG.1.review.assembly?dl=0) - *.assembly* file as exported by Juicebox Assembly Tools;
- [AGWG.rawchrom.hic](https://www.dropbox.com/s/s28y0oyefv9y4rp/AGWG.rawchrom.hic?dl=0), [AGWG.rawchrom_0.hic](https://www.dropbox.com/s/fmqra3xglgrd5vp/AGWG.rawchrom_0.hic?dl=0), [AGWG.rawchrom_30.hic](https://www.dropbox.com/s/4yz677nywxhfasy/AGWG.rawchrom_30.hic?dl=0) - Hi-C contact maps as generated by the 3D-DNA pipeline visualization module (see below), thresholded at mapping quality 1, 0 and 30, respectively (see below);
- [AGWG.rawchrom_asm.scaffold_track.txt](https://www.dropbox.com/s/bwnmosxusav5jwb/AGWG.rawchrom_asm.scaffold_track.txt?dl=0) and [AGWG.rawchrom_asm.superscaf_track.txt](https://www.dropbox.com/s/i3f8ltghb72ovvd/AGWG.rawchrom_asm.superscaf_track.txt?dl=0) - 2D annotation tracks for this step as generated by 3D-DNA visualization module (see below).

Commands used to generate the above files are as follows:
```sh
# preliminary
ln -sf AGWG.1.review.assembly AGWG.rawchrom.assembly
awk -f ${path_to_3ddna}/utils/convert-assembly-to-cprop-and-asm.awk AGWG.rawchrom.assembly
bash ${path_to_3ddna}/edit/edit-mnd-according-to-new-cprops.sh AGWG.rawchrom.cprops AGWG.mnd.txt > AGWG.mnd.rawchrom.txt
```
```sh
# generate a high-resolution Hi-C map to visualize pre-merge assembly with default mapping quality threshold (mapq>=1)
bash ${path_to_3ddna}/visualize/run-assembly-visualizer.sh -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.rawchrom.assembly AGWG.mnd.rawchrom.txt
```
```sh
# generate supplementary Hi-C maps to visualize pre-merge assembly with mapq>=0 and mapq>=30:
bash ${path_to_3ddna}/visualize/run-assembly-visualizer.sh -q 0 -m temp.AGWG.rawchrom.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.rawchrom.assembly AGWG.mnd.rawchrom.txt
bash ${path_to_3ddna}/visualize/run-assembly-visualizer.sh -q 30 -m temp.AGWG.rawchrom.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.rawchrom.assembly AGWG.mnd.rawchrom.txt
```
The progression from draft to raw chromosomal scaffolds can be interactively reviewed at http://bit.ly/2JCQMWE. This shows four contact maps, [AGWG.draft_0.hic](https://www.dropbox.com/s/ipf6sagb6j3m7fy/AGWG.draft_0.hic?dl=0), [AGWG.0_0.hic](https://www.dropbox.com/s/urtyabxyl3rywu1/AGWG.0_0.hic?dl=0), [AGWG.1_0.hic](https://www.dropbox.com/s/o76oxr6x0gpwnj0/AGWG.1_0.hic?dl=0) and [AGWG.rawchrom_0.hic](https://www.dropbox.com/s/fmqra3xglgrd5vp/AGWG.rawchrom_0.hic?dl=0), all generated using the same Hi-C dataset. The pixel intensity in the contact maps indicates how often a pair of loci colocate in the nucleus, on a scale from white to red (pure white indicates no observed contacts, whereas pure red indicates a maximum number of contacts which is indicated in the left top corner). The maps become 'cleaner', with less off-diagonal signal showing, as order and orientation of contigs changes from essentially random in the draft to locally correct in raw chromosomal scaffolds. The 18 off-diagonal peaks remaining in the raw chromosomal map are a manifestation of telomere-to-telomere and centromere-to-centromere clustering in *A. aegypti*. By default the interactive figure shows a representative interval; by scrolling readers can explore the complete, genome-wide data.

##### Step 5: Automatic pairwise alignment of nearby contigs
We then used [LASTZ](https://www.bx.psu.edu/~rsharris/lastz/) to do pairwise alignment of all contigs that fall within a sliding window of 3Mb along the raw chromosomal scaffolds. The alignment data allows one to identify long nearby stretches of high sequence identity suggestive of undercollapsed heterozygosity.
The intermediate files associated with this step include:
- [alignments.txt](https://www.dropbox.com/s/yo6cfh12amnegcv/alignments.txt?dl=0) - list of all pairwise alignments that fit the alignment score, overlap length, position and sequence identity criteria to be marked as alternative haplotypes, as generated by the 3D-DNA merge module (see below);
- [overlaps_2D_input.txt](https://www.dropbox.com/s/l62zk9r32t9dber/overlaps_2D_input.txt?dl=0) - same data in a convenience 2D annotation format, as generated by 3D-DNA merge module (see below);
- [overlaps_2D_asm.txt](https://www.dropbox.com/s/zae7wrlcp0du1xe/overlaps_2D_asm.txt?dl=0) - 2D track with positions of overlaps in genome-wide reference frame, as generated by 3D-DNA merge module (see below);
- [AGWG.draft.rawchrom_tiled.asm](https://www.dropbox.com/s/y2b6q3cy0itx3sr/AGWG.draft.rawchrom_tiled.asm?dl=0) - assembly file listing merge blocks -- chains of overlapping contigs organized based on alignment information -- as identified by the 3D-DNA merge module (see below);
- [AGWG.draft.rawchrom_tiled.assembly](https://www.dropbox.com/s/4w4swbmfglvd72i/AGWG.draft.rawchrom_tiled.assembly?dl=0) - Juicebox Assembly Tools-compatible version of the same file (see below).

For reproducing the above files, see the following commands:

```sh
# prep
awk -f ${path_to_3ddna}/utils/convert-assembly-to-cprop-and-asm.awk AGWG.rawchrom.assembly
awk -f ${path_to_3ddna}/edit/edit-fasta-according-to-new-cprops.awk AGWG.rawchrom.cprops AGWG.draft.fasta > AGWG.rawchrom.fasta
mkdir faSplit && cd faSplit && awk -f ${path_to_3ddna}/merge/split-fasta-by-cname.awk AGWG.rawchrom.cprops AGWG.rawchrom.fasta && cd ..
```
```sh
# perform pairwise alignment of all input contigs/scaffolds within set distance and filter using alignment scores, length and identity to select probable alternative haplotype sequences
bash ${path_to_3ddna}/merge/align-nearby-sequences-and-filter-overlaps.sh AGWG.rawchrom.cprops AGWG.rawchrom.asm faSplit
```
```sh
# tile based on filtered alignment results
bash ${path_to_3ddna}/merge/tile-assembly-based-on-overlaps.sh AGWG.rawchrom.cprops AGWG.rawchrom.asm overlaps_2D_input.txt
```
For interactive review of results associated with step #5 see [this interactive figure](http://bit.ly/2JIu3Z7). The figure shows the [AGWG.rawchrom_0.hic](https://www.dropbox.com/s/fmqra3xglgrd5vp/AGWG.rawchrom_0.hic?dl=0) contact map from the previous step with two different annotation sets overlaid. As before, the heatmap indicates the frequency of contact between pairs of loci as measured by the Hi-C experiment, on a scale from white to red; pure white indicates no observed contacts, whereas pure red indicates a maximum number of contacts which is indicated in the left top corner. The same scale is used for both maps by default.

On the left, 2D square annotations along the diagonal indicate the boundaries of scaffolds as ordered and oriented in Step 4, encoded in [AGWG.rawchrom_asm.scaffold_track.txt](https://www.dropbox.com/s/bwnmosxusav5jwb/AGWG.rawchrom_asm.scaffold_track.txt?dl=0) file. On the right, the same fragment of the contact map is shown with overlaid annotations of tentative overlaps as identified by pairwise LASTZ alignments ([overlaps_2D_asm.txt](https://www.dropbox.com/s/zae7wrlcp0du1xe/overlaps_2D_asm.txt?dl=0)). One can see that as a rule the regions of high sequence identity match off-diagonal stretches of high contact intensity parallel or perpendicular to the diagonal. This is brought about by the accumulation of contacts associated with Hi-C reads aligning with low mapping quality (mapq~0). The latter appear when the aligner, unable to distinguish between the haplotypes, assigns a large number of Hi-C contacts associated with true 1D proximity to one of the alternative haplotypes present in the assembly at random, effectively creating mirror images of the diagonal contact band.

The 'parallel' stretches correspond to cases when the alternative haplotypes are in the same orientation in the raw chromosomal scaffolds. The 'perpendicular' off-diagonal stretches corespond to cases where orientation of contigs as suggested by pairwise alignment does not match that suggested by the scaffolding step. These differences are reconciled in favor of the alignment signal (see [AGWG.draft.rawchrom_tiled.assembly](https://www.dropbox.com/s/4w4swbmfglvd72i/AGWG.draft.rawchrom_tiled.assembly?dl=0)).

By default the figure indicates a representative interval; by scrolling readers can explore the complete, genome-wide data and annotations.

##### Step 6: Curation of merge groups
The automatically identified merge groups from Step 5 have been curated using Assembly Tools to remove false positive alignments, remove ambiguously aligning contigs and reconcile ordering and orientation disagreements in merge groups. The files shared in association with this review are listed below:
- [AGWG.rawchrom_tiled.review.asm](https://www.dropbox.com/s/shfnht2r6zt45pf/AGWG.rawchrom_tiled.review.asm?dl=0) - the curated version of the file [AGWG.draft.rawchrom_tiled.asm](https://www.dropbox.com/s/y2b6q3cy0itx3sr/AGWG.draft.rawchrom_tiled.asm?dl=0) listing chains of overlapping contigs;
- [AGWG.rawchrom_tiled.review.assembly](https://www.dropbox.com/s/catp4s6fxt8275z/AGWG.rawchrom_tiled.review.assembly?dl=0) - Juicebox Assembly Tools-compatible version of the same file;
- [AGWG.rawchrom_tiled.review.hic](https://www.dropbox.com/s/5s9p2bd5xvdldqs/AGWG.rawchrom_tiled.review.hic?dl=0), [AGWG.rawchrom_tiled.review_0.hic](https://www.dropbox.com/s/9fa2aoa6twe2g7j/AGWG.rawchrom_tiled.review_0.hic?dl=0), [AGWG.rawchrom_tiled.review_30.hic](https://www.dropbox.com/s/s3pgf0q3faq09vx/AGWG.rawchrom_tiled.review_30.hic?dl=0) - Hi-C contact maps as generated by the 3D-DNA pipeline visualization module, thresholded at mapping quality 1, 0 and 30, respectively (see below);
- [AGWG.rawchrom_tiled.review_asm.scaffold_track.txt](https://www.dropbox.com/s/bgq4bg39tnyatts/AGWG.rawchrom_tiled.review_asm.scaffold_track.txt?dl=0), [AGWG.rawchrom_tiled.review_asm.superscaf_track.txt](https://www.dropbox.com/s/mlfaxd7174mj47f/AGWG.rawchrom_tiled.review_asm.superscaf_track.txt?dl=0) - 2D annotation tracks for this step as generated by 3D-DNA visualization module (see below).

Commands relevant to this step are as follows:
```sh
# visualize curated assembly results
awk -f ${path_to_3ddna}/utils/convert-assembly-to-cprops-and-asm.awk AGWG.rawchrom_tiled.curated.assembly
bash ${path_to_3ddna}/edit/edit-mnd-according-to-new-cprops.sh AGWG.1.cprops AGWG.mnd.txt > AGWG.mnd.1.txt
```
```sh
# order and orient scaffolds larger than 20000
bash ${path_to_3ddna}/scaffold/run-liger-assembler.sh -s 20000 AGWG.1.cprops AGWG.mnd.1.txt
```
```sh
# build high-resolution hic map with mapq >= 1 (default):
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 AGWG.1.cprops AGWG.1.asm AGWG.mnd.1.txt
```
```sh
# build supplementary high-resolution hic maps with reads mapping quality >=0 and >=30.
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -q 0 -m temp.AGWG.1.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.1.cprops AGWG.1.asm AGWG.mnd.1.txt
bash ${path_to_3ddna}/visualize/run-asm-visualizer.sh -q 30 -m temp.AGWG.1.asm_mnd.txt -r 2500000,1000000,500000,250000,100000,50000,25000,10000,5000,1000,500 -c AGWG.1.cprops AGWG.1.asm AGWG.mnd.1.txt
```
For interactive review of this step see http://bit.ly/2JLGzaq. This shows the [AGWG.rawchrom_tiled.review_0.hic](https://www.dropbox.com/s/9fa2aoa6twe2g7j/AGWG.rawchrom_tiled.review_0.hic?dl=0) with [AGWG.rawchrom_tiled.review_asm.scaffold_track.txt](https://www.dropbox.com/s/bgq4bg39tnyatts/AGWG.rawchrom_tiled.review_asm.scaffold_track.txt?dl=0) boundary annotations along the diagonal. The heatmap indicates the frequency of contact between pairs of loci as measured by the Hi-C experiment, on a scale from white to red, and the square annotations along the diagonal indicate the ordering and orientation of contigs after tiling curation. By default, the map shows the neighborhood of the same contigs that have been the focus of [interactive illustration for Step 6](http://bit.ly/2JIu3Z7); by scrolling readers can explore the complete, genome-wide data and annotations.
##### Step 7: Automatic merging of overlapping contigs

Merging of chained sequences was performed iteratively by the 3D-DNA merge module. The results are shared as follows:
- [AGWG.final.fasta.gz](https://www.dropbox.com/s/edeqbhfoj3ubsju/AGWG.final.fasta.gz?dl=0) - haploid contigs generated by 3D-DNA via merging overlapping sequences within individual merge blocks (see below);
- [overlap_qc_track_2D_asm.txt](https://www.dropbox.com/s/4r7v6xd2fckzhum/overlap_qc_track_2D_asm.txt?dl=0), [contributing_2D_asm.txt](https://www.dropbox.com/s/1ljm9zlx6kv89t6/contributing_2D_asm.txt?dl=0) and [not_contributing_2D_asm.txt](https://www.dropbox.com/s/j78i8uhficomhnf/not_contributing_2D_asm.txt?dl=0) - 2D track files that facilitate comparison of iterative and pairwise alignments as well as classify individual contig contributions to final haploid reference, as generated by the 3D-DNA merge module (see below);
- [AGWG.HiC.fasta.gz](https://www.dropbox.com/s/gltd803wcval43c/AGWG.HiC.fasta.gz?dl=0) - final chromosome-length scaffolds as generated by the 3D-DNA finalize module (see below).

To reproduce the above files see the following commands:
```sh
# create new contigs/scaffolds by merging connected components defined in the tiled assembly file
bash ${path_to_3ddna}/merge/merge-tiled-asm.sh -a AGWG.rawchrom_tiled.review_asm.scaffold_track.txt AGWG.rawchrom.cprops AGWG.rawchrom_tiled.review.asm faSplit
```
```sh
# parse iterative merge results for comparison with pairwise alignment and contig contribution classification
awk 'NR==1||$7=="255,255,0"' AGWG.rawchrom/overlap_qc_track_2D_asm.txt > contributing_2D_asm.txt
awk 'NR==1||$7=="255,255,255"' AGWG.rawchrom/overlap_qc_track_2D_asm.txt > not_contributing_2D_asm.txt
```
```sh
# finalize output from haploid contigs to haploid chromosome-length scaffolds
cp AGWG.rawchrom/AGWG.rawchrom_merged.asm AGWG.final.asm
ln -sf AGWG.rawchrom/merged_AGWG.rawchrom.fa AGWG.final.fasta
awk -f ${path_to_3ddna}/utils/generate-cprops-file.awk AGWG.final.fasta > AGWG.final.cprops
bash ${path_to_3ddna}/finalize/finalize-output.sh -l AGWG AGWG.final.cprops AGWG.final.asm AGWG.final.fasta final
ln -sf AGWG.FINAL.fasta AGWG.HiC.fasta
```
For review of iterative alignment results see  http://bit.ly/2JECh4H showing [AGWG.rawchrom_tiled.review_0.hic](https://www.dropbox.com/s/9fa2aoa6twe2g7j/AGWG.rawchrom_tiled.review_0.hic?dl=0) heatmap with three types of overlaid annotations. The heatmap indicates the frequency of contact between pairs of loci as measured by the Hi-C experiment, on a scale from white to red, with the red indicating the maximal number of contacts (indicated in the top left corner).

On the left, the diagonal annotations ([AGWG.rawchrom_tiled.review_asm.scaffold_track.txt](https://www.dropbox.com/s/bgq4bg39tnyatts/AGWG.rawchrom_tiled.review_asm.scaffold_track.txt?dl=0)) indicate the boundaries of draft contigs as ordered and oriented by the assembly at this stage.

In the center, [not_contributing_2D_asm.txt](https://www.dropbox.com/s/j78i8uhficomhnf/not_contributing_2D_asm.txt?dl=0) 2D annotation file highlights contigs that will, in their entirety, be annotated as alternate alleles. These contigs will not contribute any sequence to the 'main' haploid reference as the sequence they contain is already included as part of larger contigs. The annotations are a subset of diagonal squares from the left panel shifted up and left to match the genomic positions of overlapping sequences upstream in the assembly. As expected (and analogous to the pairwise alignment annotation in Step 5) these annotation enclose stretches of enriched contact frequency parallel to the diagonal, brought about by random assignment of short Hi-C reads with low mapping quality across haplotypes.

Overlaid on top of the map on the right is the [contributing_2D_asm.txt](https://www.dropbox.com/s/1ljm9zlx6kv89t6/contributing_2D_asm.txt?dl=0) track. This track highlights contigs that only partly overlap with the contigs upstream in the assembly. As such, these contigs will contribute sequence to the final haploid reference. The annotations represent a subset of diagonal squares from the left panel, shifted up and left to genomic positions of matching sequences upstream in the assembly. The parallel-to-diagonal stretches of enriched signal associated with the overlap are seen in the top left corner of each annotation.

By default the figure indicates a representative interval; by scrolling readers can explore the complete, genome-wide data and annotations. To explore the corresponding location of the contact map after the merge see http://bit.ly/2JA7khR (see Step 8 for details).
##### Step 8: Final NCBI submission

The Hi-C chromosomal scaffolds from Step 7 have been then polished and gap-filled using PBJelly, see [(Matthews, Dudchenko & Kingan et al., bioRxiv, 2017)] for details. Additionally, mitochondrial sequences have been extracted and gaps normalized to 100N, and the resulting sequences as well as alternate alleles [submitted to NCBI](https://www.ncbi.nlm.nih.gov/assembly/GCF_002204515.2/). Hi-C maps associated with the final output are as follows:
- [Liverpool1.hic](https://www.dropbox.com/s/w51y0agv35ocmi8/inter.hic?dl=0), [Liverpool1_30.hic](https://www.dropbox.com/s/uyfwr0nll3rbe9f/inter_30.hic?dl=0) - Hi-C contact maps created by realigning the Hi-C data from [SRR6294665](https://www.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6294665) experiment to the haploid chromosome-length AaegL5 reference, as generated by the Juicer pipeline, thresholded at mapping quality 1 and 30, respecitvely;
- [Liverpool2.hic](https://www.dropbox.com/s/fe9jyby8ap6isgv/inter.hic?dl=0), [Liverpool2_30.hic](https://www.dropbox.com/s/0l0shpkquhiio7l/inter_30.hic?dl=0) - Hi-C contact maps created by realigning the Hi-C data from [SRR6294663](https://www.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6294663) experiment to the haploid chromosome-length AaegL5 reference, as generated by the Juicer pipeline, thresholded at mapping quality 1 and 30, respectively;
- [Liverpool3.hic](https://www.dropbox.com/s/29mak0kn98q58kq/inter.hic?dl=0), [Liverpool3_30.hic](https://www.dropbox.com/s/jq83p8is9hcxezi/inter_30.hic?dl=0) - Hi-C contact maps created by realigning the Hi-C data from [SRR6294664](https://www.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6294664) experiment to the haploid chromosome-length AaegL5 reference, as generated by the Juicer pipeline, thresholded at mapping quality 1 and 30, respectively;
- [mega.hic](https://www.dropbox.com/s/33tcu35du01j78b/inter.hic?dl=0), [mega_30.hic](https://www.dropbox.com/s/qzqkl6f1g6n6eqb/inter_30.hic?dl=0) - contact maps, created by combining data from all three Hi-C experiments listed above, as generated by the [mega.sh] (https://github.com/theaidenlab/juicer/blob/1ef023d11cfda07228edf0288435e3b7015225f1/CPU/common/mega.sh) script from the Juicer suite, thresholded at mapping quality 1 and 30, respectively.

Explore the genome-wide contact map [mega.hic](https://www.dropbox.com/s/33tcu35du01j78b/inter.hic?dl=0) via http://bit.ly/2JGNmSN. This interactive heatmap indicates the frequency of contacts between pairs of loci in AaegL5 at multiple resolutions as measured by three Hi-C experiments conducted for this study, on a scale from white to red; pure white indicates no observed contacts, whereas pure red indicates a maximum number of contacts (indicated in the top left corner). By default the genome-wide map is shown; by zooming in and scrolling readers can explore the complete, genome-wide data for each of the three chromosomes.

For details on how to run the Juicer pipeline see the [Juicer GitHub page](https://github.com/theaidenlab/juicer) and [Wiki](https://github.com/theaidenlab/juicer/wiki).

### Citations and licensing
If you use AaegL5 or any of the intermediate assembly results in your work, please cite the following paper:

Matthews, B.J.,\* Dudchenko O.,\* Kingan S.,\* Koren S., Antoshechkin I., Crawford J.E., Glassford W.J., Herre M., Redmond S.N., Rose N.H., et al. (2017). *Improved Aedes aegypti mosquito reference genome assembly enables biological discovery and vector control.* bioRxiv. doi: https://doi.org/10.1101/240747.

For Hi-C assembly methods please cite:

Dudchenko, O., Batra, S.S., Omer, A.D., Nyquist, S.K., Hoeger, M., Durand, N.C., Shamim, M.S., Machol, I., Lander, E.S., Aiden, A.P., et al. (2017). *De novo assembly of the Aedes aegypti genome using Hi-C yields chromosome-length scaffolds.* Science. Apr 7; 356(6333):92-95. doi: https://doi.org/10.1126/science.aal3327. Epub 2017 Mar 23.


#### This software is distributed under The MIT License (MIT).
[3D-DNA pipeline (Dudchenko et al., Science, 2017)]: <http://science.sciencemag.org/content/356/6333/92>
[(Matthews, Dudchenko & Kingan et al., bioRxiv, 2017)]: <https://www.biorxiv.org/content/early/2017/12/29/240747>
[AaegL5]: <https://www.biorxiv.org/content/early/2017/12/29/240747>
[(Durand & Shamim et al., Cell Systems, 2016)]: <http://www.cell.com/cell-systems/abstract/S2405-4712(16)30219-8>
[(Durand & Robinson et al., Cell Systems, 2016)]: <http://www.cell.com/cell-systems/abstract/S2405-4712(15)00054-X>
[(Dudchenko et al., bioRxiv, 2018)]: <https://www.biorxiv.org/content/early/2018/01/28/254797>
[here]: <https://www.dropbox.com/s/le9cs0l6hdsqbkd/Extended_Hi-C_methods.docx?dl=0>
[3d-dna]: <https://github.com/theaidenlab/3d-dna>
[http://www.aidenlab.org/forum.html]: <http://www.aidenlab.org/forum.html>
[(Rao & Huntley et al., Cell, 2014)]: <http://www.cell.com/abstract/S0092-8674(14)01497-4>
[(Robinson et al., Cell Systems, 2018)]: <http://www.cell.com/cell-systems/fulltext/S2405-4712(18)30001-2>
